contrib/labstack/echo.v4/appsec_test.go

```diff
@@ -0,0 +1,406 @@
+			require.NotNil(t, event)
+			require.True(t, strings.Contains(event, "crs-913-120"))
+			require.True(t, strings.Contains(event, "myPathParam2"))
+			require.True(t, strings.Contains(event, "server.request.path_params"))
+		})
+
+		t.Run("wildcard", func(t *testing.T) {
+			mt := mocktracer.Start()
+			defer mt.Stop()
+			// Send a security scanner attack (according to appsec rule id crs-913-120)
+			req, err := http.NewRequest("POST", srv.URL+"/path0.0/param0/path0.1/param1/path0.2/param2/path0.3/appscan_fingerprint", nil)
+			if err != nil {
+				panic(err)
+			}
+			res, err := srv.Client().Do(req)
+			require.NoError(t, err)
+			// Check that the handler was properly called
+			b, err := io.ReadAll(res.Body)
+			require.NoError(t, err)
+			require.Equal(t, "Hello World!\n", string(b))
+			require.Equal(t, http.StatusOK, res.StatusCode)
+			// The span should contain the security event
+			finished := mt.FinishedSpans()
+			require.Len(t, finished, 1)
+			event := finished[0].Tag("_dd.appsec.json").(string)
+			require.NotNil(t, event)
+			require.True(t, strings.Contains(event, "crs-913-120"))
+			// Wildcards are not named in echo
+			require.False(t, strings.Contains(event, "myPathParam3"))
+			require.True(t, strings.Contains(event, "server.request.path_params"))
+		})
+	})
+
+	t.Run("response-status", func(t *testing.T) {
+		mt := mocktracer.Start()
+		defer mt.Stop()
+
+		req, err := http.NewRequest("POST", srv.URL+"/etc/", nil)
+		if err != nil {
+			panic(err)
+		}
+		res, err := srv.Client().Do(req)
+		require.NoError(t, err)
+		require.Equal(t, 404, res.StatusCode)
+
+		finished := mt.FinishedSpans()
+		require.Len(t, finished, 1)
+		event := finished[0].Tag("_dd.appsec.json").(string)
+		require.NotNil(t, event)
+		require.True(t, strings.Contains(event, "server.response.status"))
+		require.True(t, strings.Contains(event, "nfd-000-001"))
+	})
+
+	// Test a PHP injection attack via request parsed body
+	t.Run("SDK-body", func(t *testing.T) {
+		mt := mocktracer.Start()
+		defer mt.Stop()
+
+		req, err := http.NewRequest("POST", srv.URL+"/body", nil)
+		if err != nil {
+			panic(err)
+		}
+		res, err := srv.Client().Do(req)
+		require.NoError(t, err)
+		// Check that the handler was properly called
+		b, err := io.ReadAll(res.Body)
+		require.NoError(t, err)
+		require.Equal(t, "Hello Body!\n", string(b))
+
+		finished := mt.FinishedSpans()
+		require.Len(t, finished, 1)
+		event := finished[0].Tag("_dd.appsec.json")
+		require.NotNil(t, event)
+		require.True(t, strings.Contains(event.(string), "crs-933-130"))
+	})
+}
+
+func TestControlFlow(t *testing.T) {
+	middlewareResponseBody := "Hello Middleware"
+	middlewareResponseStatus := 433
+	handlerResponseBody := "Hello Handler"
+	handlerResponseStatus := 533
+
+	for _, tc := range []struct {
+		name        string
+		middlewares []echo.MiddlewareFunc
+		handler     func(echo.Context) error
+		test        func(t *testing.T, rec *httptest.ResponseRecorder, err error)
```

Comment:
The following tests are currently testing that the final response status is the expected one.
I'd suggest you to also start a mocktracer and check the `http.status_code` tag (when it makes sense, since some of those tests do early returns that don't reach `Middleware()` wrapper.

I wonder how we could also check the resulting appsec address value in a simple way.
Ideally, I'd also test what appsec observed, either thanks to dyngo listener for the http operation results or directly accessing your response wrapper (the answer isn't obvious 🤔)

Reply:
Added:
- mocktracer to check span tags when relevant
- general description of each test

As for monitoring addresses, another way could be to trigger a custom rule on` http.status_code` and make sure the event gets caught.

Reply:
Oops, I approved the PR before seeing this pending comment.
I think it's worth giving it a try to see what's simplest and more maintainable over time. I like your idea as it is the most blackbox testing we can imagine I guess.

Reply:
I'll stack another PR on top of this one to take care of this.